src/main/java/com/google/devtools/build/lib/analysis/RuleContext.java

```diff
@@ -2306,6 +2307,14 @@ private boolean checkRuleDependencyMandatoryProviders(
 
     private void validateDirectPrerequisite(
         Attribute attribute, ConfiguredTargetAndData prerequisite) {
+      if (RuleContextConstraintSemantics.checkForIncompatibility(prerequisite.getConfiguredTarget())
```

Comment:
Is there any reason this couldn't be put directly in `checkRuleDependencyMandatoryProviders`? It looks like none of the other checks here (like checking valid rule type) would be affected by incompatible targets?

Reply:
I could move it in there, but then I need a duplicate check in `validateDirectPrerequisiteFileTypes` I think. Basically the code that checks for extensions also needs to be skipped/tweaked.

In other words, if I move this into `checkRuleDependencyMandatoryProviders`, then the `test_dependencies_with_extensions` test would fail.

That of course works for me too. It's not like I'd be copy-pasting tens of lines of code in two places.

Reply:
Ah, got it. No worries. Looks good as is.